File: src/main/java/com/vc/medium/L873.scala
package com.vc.medium

object L873 {
  def lenLongestFibSubseq(arr: Array[Int]): Int = {
    import scala.collection.mutable
    val index = new mutable.HashMap[Int, Int]()
    val n = arr.length
    val dp = Array.ofDim[Int](n, n)
    arr.indices.foreach(i => {
      index.put(arr(i), i)
    })
    var res = 0
    (0 until n).foreach(i => {
      (0 until i).foreach(j => {
        val k = index.getOrElse(arr(i) - arr(j), -1)
        if(k >= 0 && k < j) {
          val cand = if(dp(j)(k) > 2) dp(j)(k) + 1 else 3
          dp(i)(j) = cand
          res = Math.max(res, cand)
        }
      })
    })
    if(res > 2) res else 0
  }
}

File: src/main/java/com/vc/medium/L873.java
package com.vc.medium;

import java.util.*;

class L873 {
    public int lenLongestFibSubseq(int[] A) {
        int N = A.length;
        HashMap<Integer, Integer> index = new HashMap<>();
        for (int i = 0; i < N; i++) index.put(A[i], i);

        int[][] dp = new int[N][N];
        int ans = 0;

        for (int k = 1; k < N; k++) {
            for (int j = 0; j < k; j++) {
                int i = index.getOrDefault(A[k] - A[j], -1);
                if (i >= 0 && i < j) {
                    int cand = dp[i][j] > 2 ? dp[i][j] + 1 : 3;
                    dp[j][k] = cand;
                    ans = Math.max(ans, cand);
                }
            }
        }
        return ans >= 3 ? ans : 0;
    }
}

